build-tools/packages/build-cli/src/commands/release.ts

```diff
@@ -0,0 +1,834 @@
+    cmd?: string;
+}
+
+/**
+ * Releases a release group recursively.
+ *
+ * @remarks
+ *
+ * First the release group's dependencies are checked. If any of the dependencies are also in the repo, then they're
+ * checked for the latest release version. If the dependencies have not yet been released, then the command prompts to
+ * perform the release of the dependency, then run the releae command again.
+ *
+ * This process is continued until all the dependencies have been released, after which the release group itself is
+ * released.
+ */
+export class ReleaseCommand<T extends typeof ReleaseCommand.flags>
+    extends StateMachineCommand<T>
+    implements StateHandler
+{
+    machine = UnifiedReleaseMachine.machine;
+    releaseGroup: ReleaseGroup | ReleasePackage | undefined;
+    versionScheme: VersionScheme | undefined;
+    bumpType: VersionBumpType | undefined;
+    releaseVersion: string | undefined;
+    shouldSkipChecks = false;
+    shouldCheckPolicy = true;
+    shouldCheckBranch = true;
+    shouldCheckBranchUpdate = true;
+    shouldCommit = true;
+    shouldInstall = true;
+    shouldCheckMainNextIntegrated = true;
+
+    static flags = {
+        releaseGroup: releaseGroupFlag({
+            exclusive: ["package"],
+            required: false,
+        }),
+        package: packageSelectorFlag({
+            exclusive: ["releaseGroup"],
+            required: false,
+        }),
+        bumpType: bumpTypeFlag({
+            required: false,
+        }),
+        versionScheme: versionSchemeFlag({
+            required: false,
+        }),
+        skipChecks: skipCheckFlag,
+        ...checkFlags,
+        ...StateMachineCommand.flags,
+    };
+
+    async init() {
+        await super.init();
+
+        const context = await this.getContext();
+        await this.initMachineHooks();
+        const flags = this.processedFlags;
+
+        this.releaseGroup = flags.releaseGroup ?? flags.package!;
+        this.releaseVersion = context.getVersion(this.releaseGroup);
+        this.bumpType = flags.bumpType as VersionBumpType;
+        this.versionScheme = flags.versionScheme as VersionScheme;
+
+        this.shouldSkipChecks = flags.skipChecks;
+        this.shouldCheckPolicy = flags.policyCheck && !flags.skipChecks;
+        this.shouldCheckBranch = flags.branchCheck && !flags.skipChecks;
+        this.shouldCheckMainNextIntegrated = !flags.skipChecks;
+        this.shouldCommit = flags.commit && !flags.skipChecks;
+        this.shouldInstall = flags.install && !flags.skipChecks;
+        this.shouldCheckBranchUpdate = flags.updateCheck && !flags.skipChecks;
+    }
+
+    // eslint-disable-next-line complexity
+    async handleState(state: string): Promise<boolean> {
+        const context = await this.getContext();
+        let localHandled = true;
+
+        // First handle any states that we know about. If not handled here, we pass it up to the parent handler.
+        switch (state) {
+            case "AskReleaseDetails": {
```

Comment:
Might be nice to hoist the case-inner logic out into separate methods/functions. Would make it a bit easier to read through the switch flow without having to navigate all the case-specific implementation details. Will also give nice homes for documentation.

Reply:
Yeah, I have started exploring that structure in another PR. Essentially using the switch statement as a "router" to the right method. This also should enable better testing of each state handler function, at least any that can be made pure functions.